src/snmalloc/global/bounds_checks.h

```diff
@@ -40,71 +42,48 @@ namespace snmalloc
    * `p` is the input pointer and `len` is the offset from this pointer of the
    * bounds.  `msg` is the message that will be reported along with the
    * start and end of the real object's bounds.
+   *
+   * Note that this function never returns.  We do not mark it [[NoReturn]]
+   * so as to generate better code. The function claims to return a void*,
+   * this is so it can be tail called in memcpy.  Note [[NoReturn]] prevents
+   * tailcails in GCC and Clang.
    */
-  SNMALLOC_SLOW_PATH SNMALLOC_UNUSED_FUNCTION inline void
-    report_fatal_bounds_error [[noreturn]] (
-      void* p, size_t len, const char* msg, decltype(ThreadAlloc::get())& alloc)
-  {
-    report_fatal_error(
-      "{}: {} is in allocation {}--{}, offset {} is past the end\n",
-      msg,
-      p,
-      alloc.template external_pointer<Start>(p),
-      alloc.template external_pointer<OnePastEnd>(p),
-      len);
-  }
-
-  /**
-   * The direction for a bounds check.
-   */
-  enum class CheckDirection
+  SNMALLOC_SLOW_PATH SNMALLOC_UNUSED_FUNCTION inline void*
+  report_fatal_bounds_error(const void* ptr, size_t len, const char* msg)
   {
-    /**
-     * A read bounds check, performed only when read checks are enabled.
-     */
-    Read,
```

Comment:
The check needs two pieces of information:

 - Is this a read or a write?
 - Are we checking reads or writes?

I'm not sure how you encode that in a single template parameter.  

I am using ifunc resolvers to instantiate the outer functions multiple times so that I get two versions:

 - No checks
 - Checks on writes
 - Checks on reads and writes

If it's a wrapper, then I use the underlying function for the no-checks version.  I can then switch between them based on an environment variable at process-launch time.  As long as I can still do that concisely, I'm happy.  

Making `check_bounds` take two arguments meant that I didn't need any `if constexpr` in the outer function, which simplifies my wrappers considerably because they just unconditionally forward their arguments.  My ifunc resolver is, itself, generated from a templated function that takes a `template template` parameter and instantiates it with the wrapper version to generate (read or write checks).  

Most of the libc function wrappers can be a simple template that calls `check_bounds<>` on any read/write pointer arguments and then forwards to the underlying implementation.

Reply:
So with a single template parameter I can do which has no `if constexpr` in the caller.
```C++
    // Check the bounds of the arguments.
    if (SNMALLOC_UNLIKELY(!check_bounds<Checked && ReadsChecked>(src, len)))
      return report_fatal_bounds_error(
        src, len, "memcpy with source out of bounds of heap allocation");
    if (SNMALLOC_UNLIKELY(!check_bounds<Checked>(dst, len)))
      return report_fatal_bounds_error(
        dst, len, "memcpy with destination out of bounds of heap allocation");
```
I can push the `SNMALLOC_UNLIKELY` inside the call, but it doesn't seem to bias the branches correctly.

It has to be two calls to get the good codegen, when we have error messages as we need to conditionally tail call the error message.

Reply:
Fair enough.  I can always wrap the two calls in a macro.

Reply:
So a macro would work, but you will probably need two versions one for a `void` returning function, and another that takes a type parameter for a not void function.

Reply:
One of my favourite small parts of C++ is that you can `return` an expression that evaluates to a `void` type from a function with a `void` return, so this should be possible to make generic.